rfcs/rfc-0003-discriminators.md

```diff
@@ -0,0 +1,229 @@
+
+  - Global user search (prior art: Steam)
+
+    Allow users to opt-in to a global username search / or otherwise also search through mutual members on servers, to avoid needing the discriminator altogether.
+
+- > Across Discord, almost half of all friend requests fail to connect the user with the person they wanted to match with, mostly because users enter an incorrect or invalid username due to a combination of missing discriminator and incorrect casing.
+
+  - Issues with ease of adding are addressed in the point above, however incorrect casing is irrelevant to Revolt as-is because Revolt's usernames are already case-insensitive and this will not change.
+
+    If we look at the given example:
+
+    > You meet someone IRL that you want to talk to on Discord, and they say “I’m Phibi Eight Nine Three Six!” You go home and add “phibi#8936” only to find out you added the wrong “Phibi” because your new friend’s username is actually “PhIBI#8936”.
+
+    Revolt does not permit a registration of both "PhIBI" and "phibi".
+
+- > You want to use a common name like “Mike” or “Jane” but there are already 9,999 Mikes or Janes so you’re blocked from that name altogether.
+
+  We are not restricted to just 4-digit discriminators in our implementation.
+
+- > You like to change your username a lot and get rate limited.
+
+  If we were also to add more stringent rate limits, this may be solved by also including display names.
+
+- > Your friend says they changed their name to “vernacular” but actually it’s “𝖛𝖊𝖗𝖓𝖆𝖈𝖚𝖑𝖆𝖗” and you have trouble finding them.
+
+  While this is a valid concern, I would personally put this down as the user's own issue.
+
+# Rationale and alternatives
+
+Discriminators appear to show the least disadvantages out of all of the solutions discussed so far.
+
+| Solution                                                                | Description                                                                                                           | Users have desired username | Selling disincentivized | Privacy | Usability | Difficulty |
+| ----------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------- | :-------------------------: | :---------------------: | :-----: | :-------: | :--------: |
+| **Discriminators**                                                      | Proposed in this RFC                                                                                                  |             ✅              |           ✅            |   ✅    |    ⚠️     |    Low     |
+| **Discriminators**<br/>w/ display names                                 | Show display names with greater priority to username / discriminator combinations.<br/>\* May be proposed in this RFC |             ✅              |           ✅            |   ✅    |    ⚠️     |    Low     |
```

Comment:
I just want to open another discussion thread here for some feedback, I would personally be for implementing display names along with discriminators (the current system Discord has).

Does anyone have any additional thoughts on this?

Reply:
While we aren't super restrictive on usernames in either the current or proposed system, I think this may still provide utility to users. Apart from warning users that it may be more difficult to add others, some users (myself include) just want a simple username regardless.

I think there can definitely be scenarios such as:
- A user wants to include an emoji in their name but doesn't want it in their actual username.
- A user wants to use a real life name or change their name in some capacity but want to have a different username.
- A user may want to update their name but not username since they already link in on a bunch of websites.

I think these are all reasonable asks.

> merely adding another layer to the name stack chaos

Webhooks certainly haven't helped here, but in terms of implementation this isn't a huge issue and it's communicated consistently (at least in the new client).

Reply:
I think the system of display names along with usernames with discriminators would be nice. I'm indifferent to whether the usernames allow weird unicode, however, a solution to the adding users issue could be restricting the username's characters while making display names unrestricted.

Either that, or warning users that using those characters can make friending difficult without the use of (future) things like qr codes and friend invite links may be good.

Reply:
i'm 100% for display names separate from your actual username for pretty much the exact reasons @insertish listed.

it'd be nice to be able to safely change your name for a one-off joke or something without risking losing your discriminator lol

Reply:
Hi, I just wanted to chime in with my opinion about discriminators.

Having both a username with a discriminator AND the ability to set a display name is perfect. However, as for usernames, I would like them to be as easily shareable as either through text or word of mouth.

Personally, the best system would have:
- A username, limited to only latin characters, numbers and symbols, like a dash or underscore with a 4-6 digit discriminator, since numbers are easy to remember. This would make it easy for someone to give out your username just by saying it, i.e. *"temptrash underscore four one two six"* would just be **temptrash_#4126**
- Display names would allow for unicode, emojis and whatever else the user pleases to choose.

This way we could solve the issue of shareability of usernames AND allow for the user to choose whatever display name they want to use. I've had issues in the past where a person was unable to give me their username due to it having unicode characters in it, and since they had no other separate chat app, they were unable to send me their username in that moment.

Let me know your thoughts on a system like this, either way I feel like display names + discriminators are definitely a way to go forward.

Reply:
I think discriminators are a good feature since enabling users to have similar or same user names allows people to better express themselves. Sure it may be similar to discord's one but it's a good feature, maybe we could use a % sign and a base 64 UUID at the end? I'm sure it's not that hard.

Reply:
> maybe we could use a % sign and a base 64 UUID at the end?

i'm sorry, i'm incredibly confused. what exactly are you suggesting here? because it sounds like this:
```
JackDotJS%2625f170-d05a-f65d-38ff-5d9a7a972382
```

Reply:
> > maybe we could use a % sign and a base 64 UUID at the end?
> 
> i'm sorry, i'm incredibly confused. what exactly are you suggesting here? because it sounds like this:
> 
> ```
> JackDotJS%2625f170-d05a-f65d-38ff-5d9a7a972382
> ```

I am suggesting something like

```
JackDotJS%sMHuhm9GTxuNi3hJ51287g
```

like this does: https://docs.rs/uuid-b64/latest/uuid_b64/

Reply:
but a number would be simpler tbh now that I think about it

Reply:
> > > maybe we could use a % sign and a base 64 UUID at the end?
> > 
> > 
> > i'm sorry, i'm incredibly confused. what exactly are you suggesting here? because it sounds like this:
> > ```
> > JackDotJS%2625f170-d05a-f65d-38ff-5d9a7a972382
> > ```
> 
> I am suggesting something like
> 
> ```
> JackDotJS%sMHuhm9GTxuNi3hJ51287g
> ```
> 
> like this does: https://docs.rs/uuid-b64/latest/uuid_b64/



> but a number would be simpler tbh now that I think about it

Shareability is also an important factor in this case, it would be best to stick to a, for example, four digit system.

Reply:
"hey, what's your revolt username?"
"oh yeah, it's jack-dot-js-percent-s-capital-m-capital-h-u-h-m-nine-g..."

yeah uh. i think that would be extremely overkill lmao.

Reply:
true I was just thinking of how to make unique- but yeah a number is good too

Reply:
hexadecimal also looks good though

Reply:
[Feel free to look at my comment above](https://github.com/revoltchat/rfcs/pull/3#discussion_r1222640917) - so far it is a system most have agreed upon, however, we are currently entertaining the idea of having a fully randomized discriminator system, meaning that once you create your account, you will be assigned a fully random, four digit number that will *not* be changeable. This is also to mitigate having "rare" discriminator combinations, like 0001, which would be blacklisted from the availability pool.

Reply:
Switching proposal to display names with restricted usernames, RFC will be updated shortly.